Proofreading - Guidelines for Content Creators (#9426)
mendix/docs: content/en/docs/appstore/create-content/guidelines-content-creators.md

## Introduction

> This document provides guidelines and best practices to follow before you submit content to the public Mendix Marketplace. They ensure that your content is secure, compliant with legal and data privacy standards, user friendly, consistent, and discoverable.

Reviewer: Much nicer introduction.

## General Guidelines {#general}

> These are some general guidelines to help you get started on creating Marketplace content:

Reviewer: Perhaps "Here are some …" but your suggestion is good too. Better than current text.

I wonder whether "to help you" is too weak, but they are guidelines so I think it is OK. It is certainly friendlier.

> * Set up a separate app to build and maintain your Marketplace content.

Reviewer: Good pick up of terminology used in Marketplace 👍

* Use the latest MTS or LTS of any major Studio Pro version when creating the item. If that is not possible, then use the earliest version of any major Studio Pro version.
* Ensure that the latest version of your product supports at least one Mendix LTS or MTS version.
* Your product must be complete and fully functional, as advertised upon submission.

> * In Mendix 10.21.0 and above, do not use direct associations in your module, as this can cause issues upon import.

Reviewer: Is this change necessary. I prefer the old text as it seems more like natural language. But perhaps yours is better for a non-native speaker?

> When you add content to the Marketplace, include in the component [documentation](/appstore/submit-content/#doc) any information that the end-user should know.

Reviewer: I like the idea behind this change. In English I tend to prefer 
"include x in y" rather than "include in y x" as being a bit more straightforward.

But much prefer "When you add".

## Using a GitHub Repository {#github}

> You can set up a GitHub repository to house the development content for your Marketplace component. During the submission process, you can share the repository URL as the component source on the [Package](/appstore/submit-content/#package) page.

Reviewer: ?house

> ### GitHub Repository Guidelines

Reviewer: 👍

> When setting up the GitHub repository for your component, follow these guidelines:

Reviewer: You changed "component" to "content" above. Should we be consistent with this change or is the meaning different?

* Make sure the repository name matches the name that will be used for the published Marketplace component.
* Use upper camel case to replace any spaces in the repository name. Example: *MyFirstApp*.
* Make sure the repository description states what the component does. You can use the same description in the Mendix Marketplace.
* Add a *.gitignore* file to keep your repository clean.

> ### Creating a Component Release

Reviewer: Great

To create a new component release for the Mendix Marketplace, follow these steps:

> 1. Create a new tag on the appropriate commit, on the production or the release branch in your GitHub repository.

Reviewer: I think the comma is wrong here.

2. From this tag, create a [new release in GitHub](https://help.github.com/articles/creating-releases). 

3. In this GitHub release, provide an official name and write the release notes.  

>    You can use these for the Marketplace release as well.

Reviewer: Interesting - I didn't think this would work but it did! And the linter doesn't complain. Well found.

> 4. If you add the *.mpk* file as a binary to the release tag, the Mendix Marketplace automatically syncs the *.mpk* to your new draft:

Reviewer: I would probably keep "binary file" but remove the first "file" to avoid duplication.

   {{< figure src="/attachments/appstore/submit-content/github-releases.png" >}}

> 5. Link this GitHub release to the upcoming Mendix Marketplace release by mentioning the GitHub release number in the description. For more details, see the [Package](/appstore/submit-content/#package) and [Updating Existing Marketplace Content](/appstore/submit-content/#updating) sections in *Upload to the Marketplace*.

Reviewer: 👍

When considering intellectual property (IP) in the Mendix Marketplace, follow these guidelines:

* Do not copy any text, names, or other data from Marketplace components for which you do not own the copyright.  

>   Doing so might create unnecessary ambiguity or confusion that could mislead users of those Marketplace components.

Reviewer: Not sure that these need to be on separate lines?

* Do not mention components published by others within your content. 
* Make sure your component documentation only includes intellectual property that you have created.  
  Including information or content published by others on the Marketplace could lead to your component being removed from the Marketplace.  
  This also means other components can be removed if they are using your component information or content. 
* If you come across instances of intellectual property abuse, contact the support team.

For more information, see [Apply IP Protection](/appstore/creating-content/sol-ip-protection/).

#### End-User License

You must declare whether your product includes, uses, depends on, or distributes third-party software of any kind from sources other than you or Mendix. Third-party software means any and all files, such as APIs, content, source code, or compiled libraries. 

You must include the following:
* In the product description:
  * Licensing information
  * Copyright information
* At the root directory level of the *.mpk* file:
  * The *licensing.txt* file
  * Any dependencies

>   * An OSS Readme file

Reviewer: Much clearer
(The linter will complain that there is no blank line before the list and that the indentation should be four spaces, not two, - but will sort this out when we run it on the first of every month!)

You must not use any restrictive or copyleft licenses, such as GNU GPL v2 or v3, as they limit the ability to sell commercially without disclosing source code, or place restrictions on code usage. 

If you have content which has an open-source license, such as MIT or Apache, check if the third-party libraries being used are compatible with the license. 

If you are using any [open-source licenses](https://opensource.org/licenses-old/category), make sure you have the code hosted in a public repository.  

> The Marketplace content governance team will check these cases once content is  submitted, and get back to you if they find issues. It is recommended that you do the check yourself beforehand to avoid any delays in the listing of your content.

Reviewer: 👍

> If you have created your own product specific terms and conditions, you must host them on your web page, and share the URL with the Mendix Marketplace team. The team will then attach the terms and conditions to the product listing page.

Reviewer: I don't think we need to insist it is on their web page, provided they host them somewhere.

* Delete personal data once it is no longer needed. 
* Integrate the [Privacy by Design](https://gdpr-info.eu/issues/privacy-by-design/) principles into your application by default. 
* Conclude appropriate data processing agreements with your suppliers and subcontractors, and give the customer full transparency and control over all recipients of their personal data.  
* Ensure that technical and organizational data protection measures are in place.

Let the Marketplace team know which of these options applies to your product:

* Personal data is not collected: the developer does not collect any data with the product.
* Personal data is collected, but not shared: the developer collects data with the product, but does not share it with any third-party.
* Personal data is collected and shared: the developer collects data with the product, and shares it with third-parties.

#### Export Control

Export control covers the rules and regulations mandated by different countries around developing, sharing, and delivering products. It also covers different areas and departments where the transfer of information and goods can happen across borders. 

> It is imperative to understand the main pillars that support the control system. 

Reviewer: 👍 very nice

##### Sanctioned Countries 

In the context of export control, sanctioned countries are the ones that you are not allowed to do business with.   

An embargo is a government order that restricts or limits commerce or the exchange of specific goods with a specified country. An embargo is usually created as a result of unfavorable political or economic circumstances between nations.  

> Country-specific embargoes, which dictate the prohibition of any economic activity, are currently in place for Cuba, North Korea, Iran, Syria, and the Crimea region of Ukraine. Russia and Belarus are not embargoed, but are very critical at the moment. 

Reviewer: Nice changes. I'm wondering why we start off explaining what a sanction is, and then just use the word "embargo" instead. 🤔

##### Sanctioned Parties 

> Sanctioned parties are individuals, companies, institutions, vessels, banks, etc., with whom business is either prohibited or possible, only after obtaining authorization from relevant government authorities. Sanctioned parties are published by the USA, the EU, and other national authorities. 

Reviewer: I think the comma was/is in the wrong position here.
I think the meaning is
"it is prohibited or it is possible only after obtaining authorization…"
But where the comma was/is makes it look as though it could be "prohibited only after obtaining authorization" which is obviously weird.

### Technical Guidelines  

> This section covers the technical guidelines needed to ensure that your content is free of viruses, malware, and security vulnerabilities.

Reviewer: 👍

#### Security Requirements 
 
You must abide by the following requirements for all modules, widgets, and connectors packaged into an *.mpk* file and published to the Mendix  Marketplace.

##### Virus and Malware Scan for Files

> The governance check performed by the Marketplace team includes a virus and malware scan. This means that Mendix expects that the product you submit be virus- and malware-free.  

Reviewer: ```suggestion
The governance check performed by the Marketplace team includes a virus and malware scan. This means that Mendix expects that the product you submit is virus- and malware-free.  
```

> We recommend completing a full scan using [VirusTotal](https://www.virustotal.com/gui/home/upload) prior to your submission. VirusTotal checks the file for traces of virus and malware against more than 70 virus engines, and provides a report. This ensures that you can provide the Mendix Marketplace team with a clean report while uploading your *.mpk* file.

Reviewer: Much better! 👍

  
##### File Integrity Check

> You must generate and provide an SHA-2 compatible hash (>=SHA256) of the file you upload to the Mendix Marketplace. This way, the Mendix Marketplace onboarding team can verify that the file is complete, not corrupted, and has not been tampered with.  

Reviewer: This section is also much better.
I wonder if we should explain the `>=SHA256` in words rather than in this shorthand?
for example, "using at least a 256 bit hash (SHA256)"

You can generate the hash using any inbuilt utility. Find out more about how to generate a hash for [Windows](https://www.howtohaven.com/system/how-to-hash-file-on-windows.shtml) or [Linux](https://www.techengineer.one/how-to-hash-files-in-linux/).
 
##### Vulnerability Check 

To ensure that the uploaded artifact is free of any critical and high vulnerabilities, you must scan all third party libraries and dependencies distributed with the *.mpk* file. These libraries and dependencies can be found in the **userlib** or  **widgets** folder within the *.mpk*.  

> You must ensure that the CVSS 3.0 score is less than 7.0 to pass the onboarding  criteria, per the [NIST NVD specification](https://nvd.nist.gov/vuln-metrics/cvss).  

Reviewer: A personal thing, but I don't really like using "per". I would perhaps say
```suggestion
You must ensure that the CVSS 3.0 score is less than 7.0 to pass the onboarding  criteria, as described in the [NIST NVD specification](https://nvd.nist.gov/vuln-metrics/cvss).  
```
or something like that.

> While uploading your file, you can provide the Mendix Marketplace team with a report proving the completion of the vulnerability scan and indicating that libraries are free of vulnerabilities. We recommend the [Snyk Open Source](https://snyk.io/product/open-source-security-management/) solution for scanning third-party dependencies for vulnerabilities. This is required for all new and subsequent versions of any component uploaded to the Mendix Marketplace. 

Reviewer: Good change. Perhaps even
```suggestion
When uploading your file, you must provide the Mendix Marketplace team with a report proving that you have completed a vulnerability scan and indicating that libraries are free of vulnerabilities. We recommend the [Snyk Open Source](https://snyk.io/product/open-source-security-management/) solution for scanning third-party dependencies for vulnerabilities. This is required for all new and subsequent versions of any component uploaded to the Mendix Marketplace. 
```

##### SAST Scan for OWASP top 10 vulnerabilities

You must ensure that there are no [OWASP Top 10 vulnerabilities](https://owasp.org/www-project-top-ten/) in the code.  

> Traditional code based SAST scanners cannot properly detect security issues in low-code applications. They also cannot point out any security misconfigurations. We recommend that you use a tool which can take into consideration the security best  practices outlined in [Implementing Best Practices for App Security](/howto/security/best-practices-security/), and point out security misconfigurations and bad practices.  

Reviewer: code-based?

The Mendix Marketplace contains the following products that you can use to inspect and correct your product before uploading it:  

* [Clevr ACR](https://marketplace.mendix.com/link/component/114669)
* [SIG QSM/AQM](https://www.softwareimprovementgroup.com/mendix-quality-and-security-management-powered-by-sigrid/)
* [Bizzomate](https://chromewebstore.google.com/detail/bizzomate-mendix-dev-tool/nkbokoloejkhohjlickhfkjfmbmboaof)

> * [Omnext Software Analyses Suit](https://marketplace.mendix.com/link/component/120746)

Reviewer: 👍

As the supplier ecosystem evolves, different types of services emerge. This section shares the best practices relevant for each type of offering in respect to access management and service reliability. 

##### Access Management

Follow these best practices with regards to access management:

* Issue one set of keys for each service instance.  
* Ensure that the keys provide minimal access rights for the application to perform operations on the service instance.  
  The service can choose the access level according to the service instance configuration. 
* Mendix does not store the keys which are issued to the service consumer.  
  If the consumer loses the keys, they will likely generate multiple binding keys for the same instances. Hence, ensure that the number of binding keys per service instance is not a limiting factor. 
* When the binding is deleted, make sure to delete the corresponding keys.  
  Do not reuse the keys for any other binding key request. 
* Log any operation with binding keys for auditing purposes.
* Return a 404 error code for both of the following scenarios, thus ensuring that hackers do not get any clues:
   * Absence of a resource (service instance)

>    * Access of a resource for which the accessor does not have rights 

Reviewer: Good changes, and good use of separate lines to clarify each point. 👍

##### Service Reliability

Follow these best practices with regards to service reliability:

* If the service cannot complete the provided operation within two seconds, consider using an asynchronous response mechanism. 

> * Throttle requests to avoid noisy neighborhood issues. 

Reviewer: I think the term is "noisy neighbor" rather than "noisy neighborhood"?

* Respond with a 4XX error code for invalid inputs and unmet preconditions. 
* Use an idempotent endpoint for retries in case of server errors, where the consuming app may retry invoking the service multiple times.
* Choose an appropriate scaling mechanism to handle varying loads.  

>   Horizontal scaling is preferred to vertical scaling.

Reviewer: I would keep these on the same line and use an en-dash?

* Set up appropriate tools that can monitor the service workload and raise notifications.
* Have a disaster recovery plan for your service.
* Broadcast the availability of your service and the scheduled maintenance window.

## Submitting Your Product

### Product Logo 

Display your cover image in your Mendix Marketplace listing by uploading a file with these specifications:
* *.jpg* format
* Minimum 600 x 420 px resolution
* Maximum 1 MB file size

Check the copyright branding guidelines of other companies if you use their logo as part of your image.

#### Thumbnail

The thumbnail is visible on the Mendix Marketplace homepage. It is the same image as the cover image within the listing itself.

> It is important to carefully think about the thumbnail image, since it is the first item that draws the attention of potential customers. Therefore, we recommend not using personal photos, but rather an image that clearly represents your offering.

Reviewer: Mendix recommends

* Names have to be in English. Currently, the Mendix Marketplace is international with no option to switch languages. Thus, use an English name for global rather than regional exposure.
### Product Name Taxonomy

* Do not use content type as part of the name. The Mendix Marketplace classifies products by content types and categories, which help customers with filtering and searching. Avoid using standard platform terms in the name of your product and focus on the main product features and benefits. For example, “Widget” can be substituted with “3D Viewer”.
The product name is just as important as the logo. As such, follow these rules to make sure your product's name is compliant and draws attention to your listing:

* Avoid using non-commonly-known abbreviations. It is important to draw customers' attention right from the thumbnail view, so potential customers open the listing itself to get more information. Apart from the image, the name has to be catchy.
* Use a maximum of 30 - 32 characters to ensure the name fits within the thumbnail.

> * If you use your company's name in the product name, similarly to the previous bullet point, do not abbreviate it so that customers remember your brand and pay attention to it in the future.

Reviewer: Good that you have removed this marketing advice and made the recommendations clearer 👍

>   Currently, the Mendix Marketplace is international, with no option to switch languages.

Reviewer: Might want to check this with the team given the recent focus on localization?

>   Instead, focus on the main product features and benefits. For example, “Widget” can be substituted with “3D Viewer”.

Reviewer: Do we need new lines for each sentence?

> These are some tips to help you write a proper product description:

Reviewer: ?
```suggestion
Here are some tips to help you write a proper product description:
```

>   This will help customers skim your text faster and focus on the most important parts of the text.

Reviewer: Do we need a new line here?

>   That way, if a potential customer searches using a certain word, your product may come up in the search results.

Reviewer: Do we need a new line here?

> If users or customers have additional questions about your product, or if they want to engage to discuss a challenge, they should always be able to contact you.  This can be done through the **Contact Us** form on the Mendix website, or by clicking the link to your webpage available on your product offering page. If they use the **Contact Us** form, you are notified, and are expected to respond in a timely manner.  

Reviewer: I think we need to follow up this **Contact Us** form as it isn't clear to me where this is (if it still exists).

>   You must provide a link to either of these in the product listing. 

Reviewer: This seems to contradict the **Contact Us** information, above.

* Use an English name to ensure global exposure.  
  Currently, the Mendix Marketplace is international, with no option to switch languages.

* Do not include the content type or standard platform terms in the name. The Mendix Marketplace already classifies products by content types and categories, which helps customers filter and search. Instead, focus on the main product features and benefits. For example, “Widget” can be substituted with “3D Viewer”.

* Avoid using abbreviations that are not commonly known.  

* If you use your company's name in the product name, do not abbreviate it.

### Product Description

Describe the challenges your potential customers may face daily, and how your product can help. Mendix recommends starting with a catchy one-liner, which will be used in the product's thumbnail as plain text. 

While the description itself has no character limit, the introductory one-liner is capped at 150 characters.

Here are some tips to help you write a proper product description:

* Include answers to key questions that developers have when they look for content.

> Here are some tips to help you write a proper product description:

Reviewer: You seem to have the same line twice.

* [Community Best Practices for App Performance](/refguide/community-best-practices-for-app-performance/)
* [Mendix Best Practices for Development](/refguide/dev-best-practices/)

> * [Best Practices for Creating Connectors](/appstore/creating-content/connector-guide-best-practices/)

Reviewer: 👍
